### ISSUE TITLE
`dotnet test` on .NET Framework test suite requires runtime identifier?

### ISSUE THREAD

**Issue by thomhurst** [microsoft/testfx#4469]
Is this anything to do with testing platform?

I've just added .NET Standard support to TUnit, and if I do a normal `dotnet test` it doesn't run unless I pass in a runtime identifier too.

Original issue is here: https://github.com/thomhurst/TUnit/issues/1465

dotnet test:

```
PS C:\Users\Tom\Downloads\repro1465\repro1465> dotnet test
Restore complete (1.7s)
  repro1465 failed with 1 error(s) (0.4s)
    C:\Program Files\dotnet\sdk\9.0.101\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(266,5): error NETSDK1047: Assets file 'C:\Users\Tom\Downloads\repro1465\repro1465\obj\project.assets.json' doesn't have a target for 'net472/win-x86'. Ensure that restore has run and that you have included 'net472' in the TargetFrameworks for your project. You may also need to include 'win-x86' in your project's RuntimeIdentifiers.
```

dotnet test -r win-x86:

```
PS C:\Users\Tom\Downloads\repro1465\repro1465> dotnet test -r win-x86
Restore complete (0.8s)
  repro1465 succeeded (7.2s) → bin\Debug\net472\win-x86\repro1465.exe
  repro1465 test succeeded (1.4s)

Test summary: total: 1, failed: 0, succeeded: 1, skipped: 0, duration: 1.3s
```

**Comment by Evangelink** [comment]
Looking at that issue now

**Comment by Evangelink** [comment]
@thomhurst could you please confirm the following behaviors:

1. From @AArnott sample, doing `dotnet build` fails with the same issue
2. Building from VS and then running `dotnet test --no-build` results in green run of tests

It would seems to indicate this is a msbuild/SDK issue on building (which we call when doing `dotnet test`).

**Comment by baronfel** [comment]
Got a binlog of what happens when you call MSBuild @Evangelink?

**Comment by Evangelink** [comment]
Here you go: [msbuild.binlog.txt](https://github.com/user-attachments/files/18337531/msbuild.binlog.txt)

**Comment by thomhurst** [comment]
You're right. It happens on the build. `dotnet build` causes it too:

```
PS C:\Users\Tom\Downloads\repro1465\repro1465> dotnet build
Restore complete (0.9s)
  repro1465 failed with 1 error(s) (0.4s)
    C:\Program Files\dotnet\sdk\9.0.101\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(266,5): error NETSDK1047: Assets file 'C:\Users\Tom\Downloads\repro1465\repro1465\obj\project.assets.json' doesn't have a target for 'net472/win-x86'. Ensure that restore has run and that you have included 'net472' in the TargetFrameworks for your project. You may also need to include 'win-x86' in your project's RuntimeIdentifiers.

Build failed with 1 error(s) in 6.4s
```

I managed to build in VS and then run on the CLI with --no-build like you suggested and that works fine:

```
PS C:\Users\Tom\Downloads\repro1465\repro1465> dotnet test --no-build
  repro1465 test succeeded (2.4s)

Test summary: total: 1, failed: 0, succeeded: 1, skipped: 0, duration: 1.7s
Build succeeded in 3.0s
```

**Comment by thomhurst** [comment]
I'm not well versed in the SDK and how msbuild does stuff, but is the problem that the ResolvePackageAssets task is using a win-x86 runtime identifier parameter?

![Image](https://github.com/user-attachments/assets/fe03fdc5-a4e9-4f85-8f6d-de34aa824923)

**Comment by JanKrivanek** [comment]
But it seems to go from here:

https://github.com/dotnet/sdk/blob/094238ce2d4f69fbfdf0fdf862ef714b3dd027ea/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets#L63

While not being used during Restore.

@marcpopMSFT - could you suggest who should look at this?

**Comment by marcpopMSFT** [comment]
This is specific to TUNIT correct? From the binlog, it appears that HasRuntimeOutput is true for the build but not for the restore which is why a RID is getting set for the build but not project assets json gets created during the restore. This is happening because OutputType is ending up as an exe in the build but not during the restore.

The restore downloads the tunit package which ends up setting the OutputType to exe after the restore phase in TUNIT.Engine.props.

@rainersigwald @dsplaisted any suggestions how tunit can work around this? Do we need a property they can disable the implicit RID inference for netfx apps? They could set HasRuntimeOutput to false but I don't know what other side effects that'll have.

**Comment by Evangelink** [comment]
@thomhurst I confirm that setting `<OutputType>Exe</OutputType>` directly in the project fixes the issue.

**Comment by thomhurst** [comment]
> [@thomhurst](https://github.com/thomhurst) I confirm that setting `<OutputType>Exe</OutputType>` directly in the project fixes the issue. So it's neither a `dotnet test` nor a MTP issue which is good.

Ah that's good!
I've got a `buildTransitive` props file in my NuGet package that should be setting that though: https://nuget.info/packages/TUnit.Engine/0.6.52 

But it looks like it's in an empty nested folder!? Reckon that could be the problem?

![Image](https://github.com/user-attachments/assets/d3912706-3f15-4fcb-bca1-5d2f8581f8b0)

**Comment by thomhurst** [comment]
Anyway - This is a me problem. Sorry for the red herring. Thanks for the help everyone!

**Comment by baronfel** [comment]
@thomhurst just confirming - that structure would 100% interfere with the auto-loading behaviors of NuGet. Check for double-slashes in the PackagePath for the items you're bundling?

**Comment by rainersigwald** [comment]
> The restore downloads the tunit package which ends up setting the OutputType to exe after the restore phase in TUNIT.Engine.props.

This is the core of it--I think the other test packages require that the project itself set `OutputType`. If it's set in the import, it will be set for "real" post-restore builds, but not the restore itself.

**Comment by dsplaisted** [comment]
Yes, props and targets files in packages should not be setting any properties which would modify the input to Restore, which includes the OutputType.

**Comment by Evangelink** [comment]
> > The restore downloads the tunit package which ends up setting the OutputType to exe after the restore phase in TUNIT.Engine.props.
> 
> This is the core of it--I think the other test packages require that the project itself set `OutputType`. If it's set in the import, it will be set for "real" post-restore builds, but not the restore itself.

One reason why custom MSBuild SDKs are good. When using it, we can actually set the output type in the props and everything works fine while for a normal package we have this error.

**Comment by thomhurst** [comment]
> Yes, props and targets files in packages should not be setting any properties which would modify the input to Restore, which includes the OutputType.

Weird. This has been in TUnit for ages and doesn't seem to affect .NET (Core) builds?

Only since allowing .NET Framework has it not worked.

I'd really like to be able to abstract away all the properties from users to simplify things for them.

**Comment by AArnott** [comment]
> I'd really like to be able to abstract away all the properties from users to simplify things for them.

@thomhurst for what it's worth IMO some properties are better left in the project. I wouldn't typically expect a `PackageReference` to change something so fundamental for the project as its `OutputType` property.

**Comment by rainersigwald** [comment]
> doing the same for xUnit. I do tend to think it's more expected for test because with the new platform each test app has to be an executable so it feels boilerplate.

I'd recommend that the NuGet package _validate_ the exe requirement but not try to impose it (for this reason that it does other stuff).

**Comment by thomhurst** [comment]
> > doing the same for xUnit. I do tend to think it's more expected for test because with the new platform each test app has to be an executable so it feels boilerplate.
> 
> I'd recommend that the NuGet package _validate_ the exe requirement but not try to impose it (for this reason that it does other stuff).

How would I do this?

**Comment by rainersigwald** [comment]
Something like (untested)

```xml
<Target Name="ValidateExeProject"
        BeforeTargets="BeforeBuild"
        Condition="'$(BuildingProject)' == 'true'">
    <Error Condition="'$(OutputType)' != 'Exe'"
           Text="TUnit test projects must have OutputType Exe as of version XXX" />
</Target>
```

**Comment by bradwilson** [comment]
I'm happy to do this for xUnit.net as well, but I have no idea what `'$(BuildingProject)' == 'true'` is. I can't find it documented anywhere, and when set as a condition my target never triggers.

My temptation is to use `'$(TargetFramework)' != ''` instead.

**Comment by thomhurst** [comment]
Echoing brad. Also does this go in the props file too?

**Comment by thomhurst** [comment]
Im struggling with props files on net framework ATM.

Net core I put files in build transitive and it works great I'm trying just build for net framework and not getting the results I expect.

**Comment by bradwilson** [comment]
Compare without `<OutputType>Exe</OutputType>` (missing `1.exe` for `net8.0`):

![Image](https://github.com/user-attachments/assets/b200322c-4fed-4d03-b033-411348853e3b)

With `<OutputType>Exe</OutputType>` (host `1.exe` is present):

![Image](https://github.com/user-attachments/assets/8ac4a2c2-f825-4064-a61a-fac079759d24)

**Comment by bradwilson** [comment]
v3 `1.0.2-pre.3` has the updates above.

**Comment by AArnott** [comment]
@thomhurst @bradwilson 

>  I have no idea what `'$(BuildingProject)' == 'true'` is. 

That probably excludes design-time builds. That's desirable because breaking design-time builds will badly break the VS experience including Intellisense and everything else, where you only need to get the user's attention in a single way, when they build. Breaking DT builds will create thousands of errors potentially, making the underlying cause harder to discover.

**Comment by bradwilson** [comment]
@AArnott Can you point to where this is documented? My Googling found nothing.

**Comment by AArnott** [comment]
No, I've never seen it documented. In fact I am probably wrong about the design-time build scenario. The "BeforeBuild" trigger the target uses probably avoids ever being called for design-time builds so it's not an issue.

IIRC BuildingProject is related to P2P project references. When you build with `/p:BuildProjectReferences=false`, it sets `BuildingProject` to `false` on all P2Ps, I _think_. Visual Studio builds also does this because the VS solution build manager wants to schedule the builds instead of letting the msbuild build manager do it. 

But @rainersigwald could say for sure.

**Comment by bradwilson** [comment]
@AArnott The reason I ask is that `dotnet build` does not set `$(BuildingProject)` to `true`. If I put that condition on my target, it never executes.
